### DISCUSSION TITLE
Fine tune vs training from scratch / Handwritten Text Recognition

### DISCUSSION THREAD

**Discussion by harindercnvrg** [mindee/doctr#1045] [Q&A]
Hello,

Could you please make it clear if the steps provided at [README](https://github.com/mindee/doctr/blob/main/references/recognition/README.md) are for fine tuning the already trained model or are these steps to train the model from scratch?

If these steps are for training from scratch, could you please direct us how to fine tune the model instead of training from scratch because most of the day to day users do not have the infrastructure to carry out training from scratch.

Thanks,

**Comment by felixdittrich92** [comment]
Hi @harindercnvrg 👋 , short answer: for both.. from scratch or fine tuning there is an arg --pretrained you can use if you want to initialize an already trained model (if the architecture provides or better 'has' a checkpoint) for fine tuning. Which model do you want to train ? :)

**Reply by felixdittrich92** [reply]
Btw. we plan to provide a pretrained model for each arch in 0.6.0 . So currently not all models have a pretrained version

**Reply by harindercnvrg** [reply]
Firstly, thank you for your quick response.
I am planning to create an ocr pipeline for handwritten text. I imagine the detection model will remain same and only the recognition model will have to be fine tuned. 
For recognition, I am planning to fine tune crnn_vgg16_bn.

**Reply by felixdittrich92** [reply]
Hi @harindercnvrg :wave:, you can try this but i think the results will not be really good (for handwritten) #1003 should work much better also for handwritten (if we have implemented it) . About the detection model i think you need to train this also or have you tested it on handwritten and get valuable detections ? :)

**Reply by harindercnvrg** [reply]
Hey @felixdittrich92 , thank you for your suggestion. I will definitely check it out. I was keeping my fingers crossed about the detector but I just tested it, like you said it doesn't work well with handwritten text. 
I will need to retrain or look into different architectures altogether for this task. It would be awesome if we could open up a discussion around handwritten documents and plan to add models for the same sometime in the future.

**Reply by felixdittrich92** [reply]
Hi @harindercnvrg this is already a bigger topic we have on track but yes i agree it would be good to open an feature request in issues for this :+1: (feel free to do this)

**Reply by harindercnvrg** [reply]
Hey @felixdittrich92 , I have opened a new feature request in the issues [#1049](https://github.com/mindee/doctr/issues/1049) 
Meanwhile, if you have any other suggestions on creating an OCR system(Detection and Prediction) that works well for Handwritten text, feel free to let me know.
Thank you for all the responses.

**Reply by felixdittrich92** [reply]
Hi @harindercnvrg thanks :+1: 
Mh... to detect handwritten DB or LinkNet should be able to do this (if you have enough data :sweat_smile: ) @frgfm wdyt ?

About the recognition part: i think ParSeq will be a key point with its decoding strategy to provide also handwritten but ftm you can have a look at TrOCR (https://huggingface.co/docs/transformers/model_doc/trocr) or give our MASTER implementation a try :)

I'm personally not a big fan of this architecture(TrOCR) (uses a big language model as decoder - roberta) NOTE: this model is really big and not useful on CPU

**Reply by harindercnvrg** [reply]
@felixdittrich92 this looks cool, although I feel going by character by character approach might not yield good results for handwritten text because handwritten words usually have a lot of overlap between two characters. What do you think?

**Reply by felixdittrich92** [reply]
Hi @harindercnvrg i think models which uses MultiHeadAttention should also be able to handle / learn this case for the character by character approach. Currently good working solutions for handwritten make use of a language model on top as a kind of corrector (like ABINet) same way used in Automatic Speech Recognition to improve the results. ParSeq goes a bit different way with its decoding strategy (I am still busy by reading the paper especially the decoding needs some time to understand :sweat_smile: ), but i have tested the first hf gradio demo of it there was a box you could write anything and the model was able to predict the written (mostly) correct without that it was trained on handwritten (https://huggingface.co/spaces/baudm/PARSeq-OCR   <- sec demo the box to draw is removed).

The biggest issue we have is : data !! :sweat_smile: We have currently IMGUR5K as dataset which contains handwritten but much to less data ... I have also tried to get access to IAM dataset but no chance to integrate it in docTR unfortunately

Maybe https://github.com/Belval/TextRecognitionDataGenerator  could be a solution for the recognition part to generate enough data

Or we translate https://github.com/Grzego/handwriting-generation and export it to onnx, that we can use it in our WordGenerator ( In my opinion, that would be a pretty cool but also time-consuming idea :) )

**Reply by frgfm** [reply]
Hey there :wave: 

Yes I agree that DB or LinkNet should be able to detect handwritten text :)

**Reply by harindercnvrg** [reply]
@felixdittrich92 I also looked into the IAM dataset and I agree with you. The only way forward would be generating our own handwritten dataset. I was able to find this combination a bit more helpful [CRAFT + TrOCR ](https://huggingface.co/spaces/tomofi/CRAFT-TrOCR) 
However, the problem again remains  that the recognition model was trained on IAM dataset. The model works fine for handwritten and digital text.

**Reply by felixdittrich92** [reply]
I agree that this works well but it is definitely not a performance beast :sweat_smile: As described by @frgfm this is a long term task but we have it on track :+1:

**Reply by harindercnvrg** [reply]
@felixdittrich92 you are right, it is quite slow indeed even with GPU. I appreciate all the help. This was really informative. Hopefully, I will be able to contribute more to this task moving forward so we can this repo one stop shop for all OCR needs.

**Reply by felixdittrich92** [reply]
That's the plan i have also in mind :+1: I work on the ViT implementation as backbone (for ViTSTR and ParSeq) currently. This will be a big step to have a model which can also be used for handwritten. (Detection should be fine with DB and LinkNet)

Maybe if you have something in mind about the data or how we can improve our Char- and WordGenerator feel free to open a PR or let us discuss it what or how we can improve it :)
